### ISSUE TITLE
Nondeterministic peak memory from KnapsackEvaluator.evaluate_knapsack_output(account_for_backward_pass=True)

### ISSUE THREAD

**Issue by siddddd17** [pytorch/pytorch#196512]
### 🐛 Describe the bug

`KnapsackEvaluator.evaluate_knapsack_output(..., account_for_backward_pass=True)`
returns different `peak_memory` values across Python processes for an identical
logical graph and identical evaluator inputs, including the saved-node set.

**Cause**
 
`_get_backward_memory_from_topologically_sorted_graph` in
`torch/_functorch/_activation_checkpointing/knapsack_evaluator.py` (line 57 on
`main`) linearises the graph with:
 
```python
sorted_nodes = list(reversed(list(nx.topological_sort(node_graph))))
```
 
`nx.topological_sort` is non-unique: a DAG generally admits many valid
topological orders, and the tie-breaking is affected by the graph's insertion
order. Here the relevant node insertion order is derived from iteration over a
Python set, in `graph_info_provider.py` (lines 184 and 196 on `main`):
 
```python
all_recomputable_banned_nodes_set = set(self.all_recomputable_banned_nodes)
...
candidate_graph.add_nodes_from(all_recomputable_banned_nodes_set)
```
 
`all_recomputable_banned_nodes` holds node *names* (strings), converted in
`inialize_from_graph`, so set iteration order can vary across processes under
Python's hash randomization. That changes the networkx node insertion order, and
with it the tie-breaking in the topological sort, so the simulated backward
schedule and the reported peak are not a stable function of the logical graph.

**Repro**

Eight subprocesses, each compiling the same model and evaluating one hardcoded
saved-node set. No solver is involved in the measured value.

<details>
<summary>repro_knapsack_evaluator_nondeterminism.py</summary>

```python
"""
Repro: KnapsackEvaluator.evaluate_knapsack_output(account_for_backward_pass=True)
returns different peak-memory values across processes for identical inputs.

    python repro_knapsack_evaluator_nondeterminism.py

Spawns N subprocesses. Each compiles the same tiny model, intercepts the
partitioner at its knapsack call, builds a GraphInfoProvider from the captured
joint graph, and evaluates one FIXED, hardcoded saved-node set. No solver is
involved in the measurement, so any variance is in the evaluator.

Exits non-zero if the peaks disagree.
"""

import hashlib
import subprocess
import sys

N_PROCS = 8


def child():
    import torch
    import torch.nn as nn
    import torch._functorch.config as fc
    from torch._functorch.partitioners import CustomKnapsackSolver, dp_knapsack
    from torch._functorch._activation_checkpointing.graph_info_provider import (
        GraphInfoProvider,
    )
    from torch._functorch._activation_checkpointing.knapsack_evaluator import (
        KnapsackEvaluator,
    )

    # Parallel branches give the joint DAG many valid topological orders.
    # A pure chain has essentially one and will not reproduce this.
    class Branching(nn.Module):
        def __init__(self, width=192, branches=6, depth=5):
            super().__init__()
            self.stem = nn.Linear(width, width)
            self.branches = nn.ModuleList(
                nn.Sequential(*[nn.Sequential(nn.Linear(width, width), nn.Tanh())
                                for _ in range(depth)])
                for _ in range(branches)
            )
            self.head = nn.Linear(width, width)

        def forward(self, x):
            h = self.stem(x)
            return self.head(sum(b(h) for b in self.branches)).sum()

    captured = {}

    class Probe(CustomKnapsackSolver):
        def __call__(self, memory, joint_graph, max_memory, node_info, banned):
            if "peak" not in captured:
                # fingerprint the graph so the report can show it is byte-identical
                # across processes while the reported peak is not
                names = [n.name for n in joint_graph.nodes]
                edges = sorted((u.name, v.name) for u in joint_graph.nodes for v in u.users)
                captured["hash"] = hashlib.sha256(
                    repr((names, edges)).encode()
                ).hexdigest()[:16]
                runtimes = [1.0] * len(memory)
                provider = GraphInfoProvider.inialize_from_graph(
                    joint_graph=joint_graph,
                    all_recomputable_banned_nodes=banned,
                    recorded_knapsack_input_memories=list(memory),
                    recorded_knapsack_input_runtimes=runtimes,
                )
                ev = KnapsackEvaluator(graph_info_provider=provider)
                # Fixed evaluator input: nodes chosen directly by index.
                # No solver output is used in the measured result.
                saved = list(range(0, len(memory), 2))
                recomp = [i for i in range(len(memory)) if i % 2]
                captured["n"] = len(memory)
                captured["peak"] = ev.evaluate_knapsack_output(
                    saved_nodes_idxs=saved,
                    recomputable_node_idxs=recomp,
                    account_for_backward_pass=True,
                )["peak_memory"]
            _, s, r = dp_knapsack(memory, [1.0] * len(memory), max_memory)
            return s, r

        def uuid(self):
            return None

    fc.activation_memory_budget_solver = Probe()
    fc.activation_memory_budget = 0.5
    torch._dynamo.reset()
    compiled = torch.compile(Branching(), backend="aot_eager", dynamic=False)
    compiled(torch.randn(32, 192, requires_grad=True)).backward()

    print(f"{captured['n']} {captured['hash']} {captured['peak']:.10f}")


def parent():
    import torch

    print(f"torch {torch.__version__}, {N_PROCS} processes")
    print("same model, same graph, same hardcoded saved-node set\n")

    peaks = []
    hashes = []
    for i in range(N_PROCS):
        out = subprocess.run(
            [sys.executable, __file__, "--child"], capture_output=True, text=True
        )
        if out.returncode != 0:
            print(out.stderr[-2000:])
            raise SystemExit("child failed")
        n, ghash, peak = out.stdout.strip().split()
        peaks.append(peak)
        hashes.append(ghash)
        print(f"  process {i}: n={n}  graph_sha256={ghash}  peak_memory = {peak}")

    print(f"\ndistinct graph hashes: {len(set(hashes))}")
    distinct = sorted(set(peaks))
    print(f"\ndistinct values: {len(distinct)}  ->  {distinct}")
    if len(distinct) > 1:
        print("\nFAIL: identical inputs produced different peak memory")
        return 1
    print("\nno variance in this run; try rerunning or raising N_PROCS")
    return 0


if __name__ == "__main__":
    if "--child" in sys.argv:
        child()
    else:
        raise SystemExit(parent())
```

</details>

**Observed**

The script prints a SHA-256 of `(node names, sorted edges)` from each subprocess,
so the graph structure is verifiably identical while the reported peak is not.

```
torch 2.14.0+cu130, 8 processes
same model, same graph, same hardcoded saved-node set

  process 0: n=31  graph_sha256=8a1519dcbbcadf63  peak_memory = 0.7741935484
  process 1: n=31  graph_sha256=8a1519dcbbcadf63  peak_memory = 0.7741935484
  process 2: n=31  graph_sha256=8a1519dcbbcadf63  peak_memory = 0.7741935484
  process 3: n=31  graph_sha256=8a1519dcbbcadf63  peak_memory = 0.7741935484
  process 4: n=31  graph_sha256=8a1519dcbbcadf63  peak_memory = 0.8064516129
  process 5: n=31  graph_sha256=8a1519dcbbcadf63  peak_memory = 0.8064516129
  process 6: n=31  graph_sha256=8a1519dcbbcadf63  peak_memory = 0.7419354839
  process 7: n=31  graph_sha256=8a1519dcbbcadf63  peak_memory = 0.7741935484

distinct graph hashes: 1

distinct values: 3  ->  ['0.7419354839', '0.7741935484', '0.8064516129']

FAIL: identical inputs produced different peak memory
```

Reproduced across repeated runs on a CPU-only machine (`Is CUDA available:
False`), so this is not kernel-level nondeterminism. Each individual run shows 3 distinct values; across runs, 3 distinct values have appeared: 0.7419354839, 0.7741935484, 0.8064516129.

**Expected**

One distinct value.

**Supporting evidence: fixing the hash seed removes the variance**

Eight consecutive runs with `PYTHONHASHSEED=0`:

```
distinct values: 1  ->  ['0.7419354839']
distinct values: 1  ->  ['0.7419354839']
distinct values: 1  ->  ['0.7419354839']
distinct values: 1  ->  ['0.7419354839']
distinct values: 1  ->  ['0.7419354839']
distinct values: 1  ->  ['0.7419354839']
distinct values: 1  ->  ['0.7419354839']
distinct values: 1  ->  ['0.7419354839']
```

This is consistent with set-iteration order being the source, though it is
supporting evidence rather than proof on its own, since the hash seed affects
many order-sensitive operations.

**On the reproducibility note**
 
The issue template asks that reproducibility reports be checked against
https://docs.pytorch.org/docs/stable/notes/randomness.html. This is not covered
by that document:
 
- No RNG is involved. Model weights are irrelevant to the measured value, which
  is computed from graph structure alone.
- No CUDA kernels or nondeterministic algorithm selection. The run above is on a
  CPU-only build.
- No floating-point associativity or atomics. The variance is in which schedule
  is simulated, not in how numbers are summed.
- Not a cross-release, cross-platform, or CPU-versus-GPU comparison. Identical
  binary, identical machine, consecutive runs.
The source is Python-level ordering: set iteration order determines networkx node
insertion order, which determines tie-breaking in a non-unique topological sort,
which selects the simulated backward schedule.
 
To be explicit about the claim: I am not arguing that one of these peak values is
correct and the others are wrong. Any of them may be a valid schedule's peak. The
claim is narrower, that `evaluate_knapsack_output` should be a deterministic
function of its inputs, and currently is not. Which schedule the simulator
*should* model is a separate question I would rather raise in its own issue.

**Two notes that may save reviewer time**

1. *In-process repetition does not detect this.* The hash seed is fixed for a
   process's lifetime, so calling `evaluate_knapsack_output` ten times in a loop
   returns one distinct value even on an affected build. Any regression test has
   to cross a process boundary, or construct the same logical graph with
   explicitly different node insertion orders.

2. *A smaller repro does not appear to exist.* I tried constructing
   `GraphInfoProvider` directly from literal node and edge lists, with both a
   symmetric fork-join graph and one with asymmetric branch lengths. Both were
   deterministic across 8 processes. The variance appears only via
   `inialize_from_graph` on a real FX joint graph, where the recomputable nodes
   are a subset within a larger graph context.

**Scope**

This path is opt-in: `evaluate_knapsack_output` defaults to
`account_for_backward_pass=False`. The nondeterministic behavior therefore does
not affect callers that use the default. Callers that explicitly enable
`account_for_backward_pass=True` can observe process-dependent `peak_memory`
values.

The proposed change only affects the ordering used by the backward-memory
simulation when `account_for_backward_pass=True`; it does not change the
default behavior of `evaluate_knapsack_output`.

**Suggested fix**

`nx.lexicographical_topological_sort` accepts a `key`, and
`GraphInfoProvider.graph_nodes_in_order` already holds the original FX node
order, reachable from the evaluator via `self._graph_info_provider`. Breaking
ties by original position matches the convention
`torch.fx.passes.tools_common.stable_topological_sort` uses elsewhere in the
codebase, where ready nodes are processed by original position:

```diff
- sorted_nodes = list(reversed(list(nx.topological_sort(node_graph))))
+ original_order = {
+     name: i
+     for i, name in enumerate(self._graph_info_provider.graph_nodes_in_order)
+ }
+ sorted_nodes = list(
+     reversed(
+         list(
+             nx.lexicographical_topological_sort(
+                 node_graph, key=lambda n: original_order[n]
+             )
+         )
+     )
+ )
```

With this applied locally, the repro gives:

```
torch 2.14.0+cu130, 8 processes
same model, same graph, same hardcoded saved-node set

  process 0: n=31  graph_sha256=8a1519dcbbcadf63  peak_memory = 0.5806451613
  process 1: n=31  graph_sha256=8a1519dcbbcadf63  peak_memory = 0.5806451613
  process 2: n=31  graph_sha256=8a1519dcbbcadf63  peak_memory = 0.5806451613
  process 3: n=31  graph_sha256=8a1519dcbbcadf63  peak_memory = 0.5806451613
  process 4: n=31  graph_sha256=8a1519dcbbcadf63  peak_memory = 0.5806451613
  process 5: n=31  graph_sha256=8a1519dcbbcadf63  peak_memory = 0.5806451613
  process 6: n=31  graph_sha256=8a1519dcbbcadf63  peak_memory = 0.5806451613
  process 7: n=31  graph_sha256=8a1519dcbbcadf63  peak_memory = 0.5806451613

distinct graph hashes: 1

distinct values: 1  ->  ['0.5806451613']

no variance in this run; try rerunning or raising N_PROCS
```

Verified: the repro drops from 3 distinct values to 1. 
Plain `lexicographical_topological_sort` with no key also removes the variance,
but tie-breaking by node name is arbitrary (`addmm_10` sorts before `addmm_2`),
so keying on original position seems preferable. Happy to go either way.

One thing worth confirming: the key lookup needs to be total. On the repro graph,
all 31 nodes of `node_graph` appear in the 520-entry `graph_nodes_in_order`, so
no `KeyError`. This should hold structurally, since the recomputable graph is
built from joint-graph nodes, but I have only checked it on this graph and would
appreciate confirmation that it holds on every construction path.

**Scope of this fix:** it removes process-to-process variance caused by
topological-order tie-breaking. It does *not* establish that the resulting order
is the semantically correct backward execution order. Peak memory is
schedule-dependent, and different valid topological orders give materially
different peaks on the same graph. I think that is a separate question and would
open a second issue for it rather than mix the two here.

I have this patch ready and am happy to open a PR if the approach looks right.

### AI assistance disclosure

I used Claude to assist with reviewing the relevant PyTorch source, analyzing the failure mode, and polishing the presentation of this report. I independently implemented and ran the reproduction, verified the cross-process behavior and graph fingerprints, reviewed the proposed fix, and take responsibility for the technical claims and results reported here.

### Versions

❯ curl -sL https://raw.githubusercontent.com/pytorch/pytorch/main/torch/utils/collect_env.py | python
Collecting environment information...
PyTorch version: 2.14.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCm SDK used to build PyTorch: N/A
HIP used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: Could not collect
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.12.3 (main, Jun 19 2026, 12:46:00) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-138-generic-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           39 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  4
On-line CPU(s) list:                     0-3
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Core(TM) i5-5350U CPU @ 1.80GHz
CPU family:                              6
Model:                                   61
Thread(s) per core:                      2
Core(s) per socket:                      2
Socket(s):                               1
Stepping:                                4
CPU(s) scaling MHz:                      84%
CPU max MHz:                             2900.0000
CPU min MHz:                             500.0000
BogoMIPS:                                3600.03
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb pti ssbd ibrs ibpb stibp tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap intel_pt xsaveopt dtherm ida arat pln pts vnmi md_clear flush_l1d ibpb_exit_to_user
Virtualization:                          VT-x
L1d cache:                               64 KiB (2 instances)
L1i cache:                               64 KiB (2 instances)
L2 cache:                                512 KiB (2 instances)
L3 cache:                                3 MiB (1 instance)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-3
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             KVM: Mitigation: VMX disabled
Vulnerability L1tf:                      Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds:                       Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown:                  Mitigation; PTI
Vulnerability Mmio stale data:           Unknown: No mitigations
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Retpolines; IBPB conditional; IBRS_FW; STIBP conditional; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Mitigation; Microcode
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] ament-flake8==0.17.5
[pip3] numpy==2.5.3
[pip3] nvidia-cublas==13.1.1.3
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.24.0.43
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu13==2.30.7
[pip3] nvidia-nvjitlink==13.4.52
[pip3] nvidia-nvtx==13.0.85
[pip3] torch==2.14.0
[pip3] triton==3.8.0
[conda] No relevant packages

cc @soulitzer @mruberry @kurtamohler @chauhang @penguinwu
